### PR TITLE
Decouple futures completion from dispatchers

### PR DESCRIPTION
**Pull request by yglukhov**
This sudden idea is pretty trivial. This PR restores old future completion behavior that doesn't rely on selectors, and fixes the recursion problem differently. `callSoon` is now reentrant and launches callback processing only for the first callback. The benefits are that futures are now self sufficient and don't need a completion proc to complete properly. This also simplifies the implementation of selectors as they now don't need to care about pending futures. I will clean up the selectors code in this regards as soon as this (and probably upcoming) is merged. Please review.
@Araq, @dom96, @zielmicha, @cheatfate.

### CODE REVIEW

**Comment by yglukhov** [comment]
@endragor, rumors say you could help test this. Could you please? =)

**Comment by zielmicha** [comment]
Looks good to me.
- it's not obvious that the code is correct (e.g. a function is not called more than once), maybe add some comments explaining why?
- I think `getCallSoonProc` should be removed, not deprecated

**Comment by yglukhov** [comment]
@zielmicha, ok, added a comment. Hopefully it clears the things up a bit.

I don't want to remove those procs to not introduce conflicts with dom's upcoming merge. All of those can later be removed completely from selectors along with pending callbacks logic. Though I could do that if @dom96 doesn't mind.

**Comment by cheatfate** [comment]
I'm just curious, for what reason the name of this PR is "Decouple futures completion from selectors".
Does `selectors` completing futures? Nope.
Does `selectors` calls callbacks? Nope.
Does `selectors' knows anything about futures? Nope.

**Comment by yglukhov** [comment]
Haha, good point, that should be "dispatchers" obviously.

**Comment by Araq** [comment]
@dom96 Please give some feedback.

File: lib/pure/asyncfutures.nim [hunk 1]
```diff
@@ -32,27 +32,38 @@ type
...
-  if callSoonProc.isNil:
-    # Loop not initialized yet. Call the function directly to allow setup code to use futures.
-    cbproc()
-  else:
-    callSoonProc(cbproc)
+  ## The `cbproc` will be executed immediately unless `callSoon`
+  ## is called recursively. In such case `cbproc` will be called
+  ## as soon as currently handled `cbproc` is complete.
+  if pendingCallbacks.isNil:
+      pendingCallbacks = newSeqOfCap[CallbackFunc](8)
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]:
```suggestion
    pendingCallbacks = newSeqOfCap[CallbackFunc](8)
```

File: lib/pure/asyncfutures.nim [hunk 2]
```diff
@@ -32,27 +32,38 @@ type
...
+
+  # More to think about. pendingCallbacks may grow as long as how many calls to
+  # callSoon are made during a single future completion cascade. That is unlikely
+  # to cause problems, but could be fixed by shrinking pendingCallbacks upon
+  # reaching some threshold.
+
+  if pendingCallbacks.len == 1:
+    var i = 0
+    while i < pendingCallbacks.len:
+      pendingCallbacks[i]()
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: Could we end up in a situation where this line raises, causing `pendingCallbacks.len` to be something higher than "1" and thus the above branch fails forever, leading to all callbacks not completing?

**Comment by Araq** [comment]
Ok, this means it's complex, **will**  break code and seems to be another case of moving the problem around without solving much. IMHO this means this should be closed.

**Comment by dom96** [comment]
I think we should discuss it instead of just closing it and forgetting about it.

**Comment by yglukhov** [comment]
It's been a while, and I can see some downsides to this approach. Namely:
- As @dom96 noted, exceptions raised in callbacks are less deterministic (even if the reentrancy logic works), and we can end up in a situation when calling `callSoon(myCB)` raises an irrelevant exception.
- Infinitely mutually recursive callbacks will block the dispatcher from working.

IMO these issues have to be sorted, otherwise merging this pr is just moving problems around, as @Araq says.

Therefore, I suggest closing this. While it still remains a useful chunk of code, it is not sufficient on its own.

**Comment by dom96** [comment]
Alright. Closing.

**Pull request closed without merging**